core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -282,8 +266,7 @@ class ProducerIdMapping(val config: LogConfig,
 
   def truncateAndReload(logEndOffset: Long, currentTime: Long) {
     truncateSnapshotFiles(logEndOffset)
-    def checkNotExpired = (producerIdEntry: ProducerIdEntry) => { isEntryValid(currentTime, producerIdEntry) }
-    loadFromSnapshot(logEndOffset, checkNotExpired)
+    loadFromSnapshot(logEndOffset, producerIdEntry => isEntryValid(currentTime, producerIdEntry))
```

Comment:
On the first point, snapshot removal is handled above in `truncateSnapshotFiles`. On the second point, I think we should not update the map end offset until we have finished rebuilding the mapping from the offset. Recall that there will usually be some entries prior to the log end offset which have to inserted into the mapping by scanning (see `Log.buildAndRecoverPidMap`).